syncer/syncer.go

```diff
@@ -231,38 +247,158 @@ func (s *Syncer) processBlock(
+// addBlockIndicies appends a range of indicies (from
+// startIndex to endIndex, inclusive) to the
+// blockIndicies channel. When all indicies are added,
+// the channel is closed.
+func (s *Syncer) addBlockIndicies(
 	ctx context.Context,
+	blockIndicies chan int64,
+	startIndex int64,
 	endIndex int64,
 ) error {
-	blockMap, err := s.fetcher.BlockRange(ctx, s.network, s.nextIndex, endIndex)
-	if err != nil {
-		return err
+	defer close(blockIndicies)
+	i := startIndex
+	for i <= endIndex {
+		// Don't load if we already have a healthy backlog.
+		if uint64(len(blockIndicies)) > s.concurrency {
+			time.Sleep(defaultFetchSleep)
+			continue
+		}
+
+		select {
+		case blockIndicies <- i:
+			i++
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+	return nil
+}
+
+// fetchChannelBlocks fetches blocks from a
+// channel with retries until there are no
+// more blocks in the channel or there is an
+// error.
+func (s *Syncer) fetchChannelBlocks(
+	ctx context.Context,
+	network *types.NetworkIdentifier,
+	blockIndicies chan int64,
+	results chan *types.Block,
+) error {
+	for b := range blockIndicies {
+		block, err := s.helper.Block(
+			ctx,
+			network,
+			&types.PartialBlockIdentifier{
+				Index: &b,
+			},
+		)
+		if err != nil {
+			return err
+		}
+
+		select {
+		case results <- block:
+		case <-ctx.Done():
+			return ctx.Err()
+		}
 	}
 
+	return nil
+}
+
+// processBlocks is invoked whenever a new block is fetched. It attempts
+// to process as many blocks as possible.
+func (s *Syncer) processBlocks(
+	ctx context.Context,
+	cache map[int64]*types.Block,
+	endIndex int64,
+) error {
+	// We need to determine if we are in a reorg
+	// so that we can force blocks to be fetched
+	// if they don't exist in the cache.
+	reorgStart := int64(-1)
+
 	for s.nextIndex <= endIndex {
-		block, ok := blockMap[s.nextIndex]
-		if !ok { // could happen in a reorg
-			block, err = s.fetcher.BlockRetry(
+		block, exists := cache[s.nextIndex]
+		if !exists {
+			// Wait for more blocks if we aren't
+			// in a reorg.
+			if reorgStart < s.nextIndex {
+				break
+			}
+
+			// Fetch the nextIndex if we are
+			// in a re-org.
+			newBlock, err := s.helper.Block(
 				ctx,
 				s.network,
 				&types.PartialBlockIdentifier{
 					Index: &s.nextIndex,
 				},
 			)
 			if err != nil {
-				return err
+				return fmt.Errorf("%w: unable to fetch block %d", err, s.nextIndex)
 			}
+
+			block = newBlock
 		} else {
 			// Anytime we re-fetch an index, we
 			// will need to make another call to the node
 			// as it is likely in a reorg.
-			delete(blockMap, s.nextIndex)
+			delete(cache, s.nextIndex)
 		}
 
-		if err = s.processBlock(ctx, block); err != nil {
-			return err
+		lastProcessed := s.nextIndex
+		if err := s.processBlock(ctx, block); err != nil {
+			return fmt.Errorf("%w: unable to process block", err)
 		}
+
+		if s.nextIndex < lastProcessed && reorgStart == -1 {
+			reorgStart = lastProcessed
+		}
+	}
+
+	return nil
+}
+
```

Comment:
How about adding a comment here? something like
```suggestion
syncRange fetches and processes a range of blocks concurrently starting from the next index to the given endIndex inclusively 
```